File: trunk/compact-net/CompactNetTupleEngine/Term.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace it.mintlab.mobilenet.tupleengine
{
    public abstract class Term
    {
        internal abstract bool unify(Term t, List<Var> vars1, List<Var> vars2);

        public abstract object getRefObject();

        public bool unify(Term t, Dictionary<string, Term> varMap)
        {
            List<Var> vars1 = new List<Var>();
            List<Var> vars2 = new List<Var>();
            lock (this)
            {
                bool result = unify(t, vars1, vars2);
                foreach (Var v in vars1)
                {
                    varMap.Add(v.getName(), v.getValue());
                    v.deunify();
                }
                return result;
            }
        }

        public bool match(Term t)
        {
            List<Var> vars1 = new List<Var>();
            List<Var> vars2 = new List<Var>();
            lock (this)
            {
                bool result = unify(t, vars1, vars2);
                foreach (Var v in vars1) v.deunify();
                foreach (Var v in vars2) v.deunify();
                return result;
            }
        }

        public override bool Equals(object obj)
        {
            return this.match((Term)obj);
        }
    }
}

File: java/JavaTupleEngine/src/it/mintlab/desktopjava/tupleengine/Term.java
package it.mintlab.desktopjava.tupleengine;

	public abstract class Term
	{
		public abstract boolean unify(Term t, java.util.ArrayList<Var> vars1, java.util.ArrayList<Var> vars2);
		public abstract Object getRefObject();

		public final boolean unify(Term t, java.util.HashMap<String, Term> varMap)
		{
			java.util.ArrayList<Var> vars1 = new java.util.ArrayList<Var>();
			java.util.ArrayList<Var> vars2 = new java.util.ArrayList<Var>();
			synchronized (this)
			{
				boolean result = unify(t, vars1, vars2);
				for (Var v : vars1)
				{
					varMap.put(v.getName(), v.getValue());
					v.deunify();
				}
				return result;
			}
		}

		public final boolean match(Term t)
		{
			java.util.ArrayList<Var> vars1 = new java.util.ArrayList<Var>();
			java.util.ArrayList<Var> vars2 = new java.util.ArrayList<Var>();
			synchronized (this)
			{
				boolean result = unify(t, vars1, vars2);
				for (Var v : vars1)
				{
					v.deunify();
				}
				for (Var v : vars2)
				{
					v.deunify();
				}
				return result;
			}
		}

		@Override
		public boolean equals(Object obj)
		{
			return this.match((Term)obj);
		}
	}